torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,399 @@
+  size_t offset = 0;
+  for (const auto& item : managed_live_ranges) {
+    auto ulvr = item.first;
+    auto size = item.second;
+    auto id = item.first.id;
+    auto aligned_size = MemoryPlanner::computeAlignedTensorSize(size);
+    allocations.push_back({ulvr, {offset, aligned_size}});
+    offset += aligned_size;
+  }
+  return allocations;
+}
+
+c10::optional<size_t> computeStorageSize(const Value& value) {
+  auto node_header = getHeader(value.node());
+  auto ttp = value.type()->cast<TensorType>();
+  if (!ttp) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " isn't a tensortype: ",
+        *value.type());
+    return c10::nullopt;
+  }
+  if (!ttp->scalarType().has_value()) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " was profiled but didn't have a scalar type: ",
+        *ttp);
+    return c10::nullopt;
+  }
+  if (!ttp->sizes().concrete_sizes().has_value()) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " was profiled but doesn't have sizes: ",
+        *ttp);
+    return c10::nullopt;
+  }
+
+  auto scalar_type = ttp->scalarType();
+  if (!scalar_type.has_value()) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " doesn't have a scalar type: ",
+        *ttp);
+    return c10::nullopt;
+  }
+
+  auto element_size = c10::elementSize(scalar_type.value());
+  // TODO: when does this fail? answer: in place mutation
+  auto numel = ttp->numel();
+  if (!numel.has_value()) {
+    TORCH_WARN(
+        node_header, "; ", value.debugName(), " doesn't have numel: ", *ttp);
+    return c10::nullopt;
+  }
+
+  return numel.value() * element_size;
+}
+
+Node* insertSlabNode(
+    const std::shared_ptr<Graph>& graph,
+    size_t total_size,
+    c10::optional<at::Device> device_type = c10::nullopt) {
+  auto slab = graph->create(prim::AllocateSlab, 1);
+  slab->output()->setType(StorageType::get());
+  slab->i_(attr::total_size, (int64_t)total_size);
+
+  if (device_type || (device_type = jit::tensorexpr::pickDeviceType(graph))) {
+    slab->i_(attr::device_type, static_cast<int8_t>(device_type->type()));
+  } else {
+    slab->i_(attr::device_type, static_cast<int8_t>(at::kCPU));
+  }
+  slab->insertBefore(graph->nodes().front());
+  return slab;
+}
+
+void insertAllocNodes(
+    const std::shared_ptr<Graph>& graph,
+    size_t total_size,
+    std::vector<std::pair<const Value*, MemRegion>>& managed_value_allocations,
+    c10::optional<at::Device> device_type = c10::nullopt) {
+  auto slab = insertSlabNode(graph, total_size, device_type);
+  auto release_slab = graph->create(prim::ReleaseSlab, 0);
+  release_slab->addInput(slab->output());
+
+  for (auto& item : managed_value_allocations) {
+    auto val = item.first;
+    auto region = item.second;
+
+    // const_cast fishy?
+    Node* node = const_cast<Node*>(val->node());
```

Comment:
Lmao static runtime does a bunch of const casting from jit code base and we do casting back... tbh I find const mostly useless especially in the context of Node/Value where you can still mutate it.. it also makes a lot of things in alias analysis annoying wrt incovariant containers